pkg/executor/mem_reader.go

```diff
@@ -859,6 +861,115 @@ func (iter *defaultRowsIter) Next() ([]types.Datum, error) {
 	return nil, nil
 }
 
+// memRowsIterForTable combine a kv.Iterator and a kv decoder to get a memRowsIter.
+type memRowsIterForTable struct {
+	kvIter   *txnMemBufferIter // txnMemBufferIter is the kv.Iterator
+	cd       *rowcodec.ChunkDecoder
+	chk      *chunk.Chunk
+	datumRow []types.Datum
+	*memTableReader
+}
+
+func (iter *memRowsIterForTable) Next() ([]types.Datum, error) {
+	curr := iter.kvIter
+	var ret []types.Datum
+	for curr.Valid() {
+		key := curr.Key()
+		value := curr.Value()
+		if err := curr.Next(); err != nil {
+			return nil, errors.Trace(err)
+		}
```

Comment:
Will this cause some problems? For example, the current key and value meet the requirements, but `curr.Next()` causing an error to be returned.

Reply:
OK, let me explain why the code looks like this:

The signature for Valid() method of kv.Iterator is Valid() bool
It's not expected to return an error.

txnMemBufferIter implements kv.Iterator, but in its Valid() implementation, error might happen.
So it record error in its iter.err field.

And when Next() is called, the error would be throw out.

So here it is basically equvalent to

```
if err := iter.Valid(); err != nil {
    return err
}
```

And once error happens, the error is throw to the top most of the caller, and I guess finally to the   mysql client.

Reply:
So when `Valid()` meets error, it also return true which feel strange to me.

Reply:
I think the current kv.Iterator API is somehow better than https://pkg.go.dev/database/sql#Rows.Next

It has a `Err()` method, and `Next()` return false for two reasons:
1. finish
2. error

While our `Valid()` just serve one purpose: 
1. finish

Our kv.Iterator do not provide the `Err()`, (because it's really common to forget check that after the for-loop). 
We return the error in the `Next()` which is much easier for the user to notice it.